From 8f73ac06bd98e606c03b351747959eee055b35ef Mon Sep 17 00:00:00 2001
From: Naveen M K <naveen521kk@gmail.com>
Date: Sat, 5 Aug 2023 14:10:27 +0530
Subject: [PATCH 067/N] dynload_win: add env var for reverting the legacy DLL
 loading paths

PYTHONLEGACYWINDOWSDLLLOADING makes Python use LOAD_WITH_ALTERED_SEARCH_PATH
again when loading DLLs. We currently need this with an uninstalled Python
and for code which does not use add_dll_directory(), or doesn't have a way
to use it.

Ideally it should be removed at some point.

Signed-off-by: Naveen M K <naveen521kk@gmail.com>
---
 Lib/test/__main__.py                   |  8 ++++++++
 Python/dynload_win.c                   | 17 ++++++++++++++---
 Tools/build/check_extension_modules.py |  4 ++++
 mingw_smoketests.py                    |  3 +++
 4 files changed, 29 insertions(+), 3 deletions(-)

diff --git a/Lib/test/__main__.py b/Lib/test/__main__.py
index 82b50ad..25bac39 100644
--- a/Lib/test/__main__.py
+++ b/Lib/test/__main__.py
@@ -1,2 +1,10 @@
+import os
+import sys
+
 from test.libregrtest.main import main
+
+if sys.platform == "win32":
+    # Enable DLL loading from PATH.
+    os.environ["PYTHONLEGACYWINDOWSDLLLOADING"] = "1"
+
 main(_add_python_opts=True)
diff --git a/Python/dynload_win.c b/Python/dynload_win.c
index 6324063..4f64dd0 100644
--- a/Python/dynload_win.c
+++ b/Python/dynload_win.c
@@ -6,6 +6,7 @@
 #include "pycore_importdl.h"      // dl_funcptr
 #include "pycore_interp.h"        // _PyInterpreterState_GetConfig()
 #include "pycore_pystate.h"       // _PyInterpreterState_GET()
+#include "pycore_initconfig.h"
 
 #include "patchlevel.h"           // PY_MAJOR_VERSION
 #include <windows.h>
@@ -212,6 +213,18 @@ dl_funcptr _PyImport_FindSharedFuncptrWindows(const char *prefix,
     _Py_CheckPython3();
 #endif /* Py_ENABLE_SHARED */
 
+    int use_legacy = 0;
+    DWORD load_library_flags = 0;
+
+    _Py_get_env_flag(1, &use_legacy, "PYTHONLEGACYWINDOWSDLLLOADING");
+
+    if (use_legacy) {
+        load_library_flags = LOAD_WITH_ALTERED_SEARCH_PATH;
+    } else {
+        load_library_flags = LOAD_LIBRARY_SEARCH_DEFAULT_DIRS |
+                              LOAD_LIBRARY_SEARCH_DLL_LOAD_DIR;
+    }
+
     wchar_t *wpathname = PyUnicode_AsWideCharString(pathname, NULL);
     if (wpathname == NULL)
         return NULL;
@@ -232,9 +245,7 @@ dl_funcptr _PyImport_FindSharedFuncptrWindows(const char *prefix,
            AddDllDirectory function. We add SEARCH_DLL_LOAD_DIR to
            ensure DLLs adjacent to the PYD are preferred. */
         Py_BEGIN_ALLOW_THREADS
-        hDLL = LoadLibraryExW(wpathname, NULL,
-                              LOAD_LIBRARY_SEARCH_DEFAULT_DIRS |
-                              LOAD_LIBRARY_SEARCH_DLL_LOAD_DIR);
+        hDLL = LoadLibraryExW(wpathname, NULL, load_library_flags);
         Py_END_ALLOW_THREADS
         PyMem_Free(wpathname);
 
diff --git a/Tools/build/check_extension_modules.py b/Tools/build/check_extension_modules.py
index 668db8d..fb5c69b 100644
--- a/Tools/build/check_extension_modules.py
+++ b/Tools/build/check_extension_modules.py
@@ -69,6 +69,10 @@ WINDOWS_MODULES = {
     "winsound",
 }
 
+if sys.platform == "win32":
+    # Enable DLL loading from PATH.
+    os.environ["PYTHONLEGACYWINDOWSDLLLOADING"] = "1"
+
 
 logger = logging.getLogger(__name__)
 
diff --git a/mingw_smoketests.py b/mingw_smoketests.py
index 7aff2d4..bbaaaeb 100644
--- a/mingw_smoketests.py
+++ b/mingw_smoketests.py
@@ -35,6 +35,9 @@ if sys._use_alt_sep:
 else:
     SEP = "\\"
 
+if sysconfig.is_python_build():
+    os.environ["PYTHONLEGACYWINDOWSDLLLOADING"] = "1"
+
 _UCRT = 'ucrt' in sysconfig.get_platform()
 
 
